pstoreds/addr_book_gc.go

```diff
@@ -0,0 +1,299 @@
+package pstoreds
+
+import (
+	"fmt"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+	query "github.com/ipfs/go-datastore/query"
+	errors "github.com/pkg/errors"
+
+	peer "github.com/libp2p/go-libp2p-peer"
+	pb "github.com/libp2p/go-libp2p-peerstore/pb"
+
+	b32 "github.com/whyrusleeping/base32"
+)
+
+var (
+	// GC lookahead entries are stored in keys with pattern:
+	// /peers/gc/addrs/<unix timestamp of next visit>/<peer ID b32> => nil
+	gcLookaheadBase = ds.NewKey("/peers/gc/addrs")
+	// in GC routines, how many operations do we place in a batch before it's committed.
+	gcOpsPerBatch = 20
+	// queries
+	purgeQuery             = query.Query{Prefix: gcLookaheadBase.String(), KeysOnly: true}
```

Comment:
The issue is that we can't *rely* on any ordering guarantees. We should fix go-ds-badger and go-ds-leveldb to type-assert the order type and just "do the right thing" (I thought we already did that...). Really, I believe this was the intended use for these `Order` types.

> For example, Badger supports traversing forwards and backwards, setting prefetch sizes, seeking to a specific point, etc. none of which we support as a result of the generalisation.

This is just something you have to deal with when you add a layer of abstraction (and we do *need* that layer). However, we should be exposing as much functionality as possible (e.g., add a `Seek` method that we can specialize for badger).

Note: I say we *need* that abstraction because, without it, we wouldn't be able to explore multiple datastores without making massive changes throughout the code base. Unfortunately, the datastore abstraction isn't really a *database* abstraction so using it when we need a database tends to be frustrating. We've been discussing adopting some kind of graph or even an SQL database in go-ipfs for cases when we really do want a database.

Reply:
After a little going back and forth because we find out that the composing `Orders` was broken (it was not hierarchical), this is resolved upstream and we now explicitly use `OrderByKey`.